### PR TITLE
pallet-emergency-para-xcm: pause xcm message processing when chain is stalled

### PR DESCRIPTION
**Pull request by elfedy**
Adds a pallet to automatically pause the execution (but not the en-queuing) of xcm messages whenever the chain has not produced a block for a given amount of relay chain blocks. There's the option to give an origin the ability to perform a runtime update whenever this "paused" mode is entered.

### CODE REVIEW

File: pallets/emergency-para-xcm/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+	#[pallet::storage]
+	/// Whether incoming XCM is enabled or paused
+	pub type Mode<T: Config> = StorageValue<_, XcmMode, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_: BlockNumberFor<T>) -> Weight {
+			// Account for 1 read and 1 write to `Mode`
+			T::DbWeight::get().reads_writes(1, 1)
```
Comment (RomarQ) [inline pallets/emergency-para-xcm/src/lib.rs:133]: This does not seem to be necessary

File: pallets/emergency-para-xcm/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+			Mode::<T>::set(XcmMode::Normal);
+			<Pallet<T>>::deposit_event(Event::NormalXcmOperationResumed);
+
+			Ok(().into())
+		}
+
+		/// Authorize a runtime upgrade. Only callable in `Paused` mode
+		#[pallet::call_index(1)]
+		#[pallet::weight((T::DbWeight::get().read, DispatchClass::Operational))]
```
Comment (RomarQ) [inline pallets/emergency-para-xcm/src/lib.rs:158]: This extrinsic also writes in `authorize_upgrade`, we should account for it even if this is only executed by authorised accounts.

File: pallets/emergency-para-xcm/src/lib.rs (hunk 1 above)
Comment (librelois) [reply pallets/emergency-para-xcm/src/lib.rs:133]: It is necessary to account for `check_associated_relay_number` stuff, @elfedy can you add a comment?

File: pallets/emergency-para-xcm/src/lib.rs (hunk 2 above)
Comment (librelois) [inline pallets/emergency-para-xcm/src/lib.rs:158]:
```suggestion
		#[pallet::weight((T::SystemWeightInfo::authorize_upgrade().saturating_add(T::DbWeight::get().read), DispatchClass::Operational))]
```

Comment (librelois) [reply pallets/emergency-para-xcm/src/lib.rs:158]: @elfedy You just need to add `authorize_upgrade` weights, cf. my suggestion

**Review by librelois — APPROVED** [review]
You just need to change one weight hint, but overall it's good

**Review by RomarQ — APPROVED** [review]

**Review by Agusrodri — APPROVED** [review]

**Pull request merged**